data_structures/linked_list.ts

```diff
@@ -0,0 +1,475 @@
+   *
+   * assertEquals(list.shift(), "a");
+   * assertEquals(list.shift(), "b");
+   * assertEquals(list.shift(), "c");
+   *
+   * assertEquals(list.length, 1);
+   * ```
+   */
+  shift(): T | undefined {
+    if (!this.#head) {
+      return undefined;
+    }
+
+    const node = this.#head;
+    this.#head = node.next;
+
+    if (this.#head) {
+      this.#head.prev = null;
+    } else {
+      this.#tail = null;
+    }
+
+    --this.#length;
+    return node.value;
+  }
+
+  /**
+   * Inserts the given value to the specified index.
+   *
+   * The complexity of this operation is linear O(n).
+   *
+   * @param value The value to insert.
+   * @param index The position to insert.
+   * @throws If `index < 0` or `index > length`
+   *
+   * @example Usage
+   * ```ts
+   * import { LinkedList } from "@std/data-structures";
+   * import { assertEquals } from "@std/assert";
+   *
+   * const list = LinkedList.from([2, 4, 6, 8]);
+   * assertEquals([...list], [2, 4, 6, 8]);
+   *
+   * list.insert(1, 0);
+   * assertEquals([...list], [1, 2, 4, 6, 8]);
+   *
+   * list.insert(3, 2);
+   * assertEquals([...list], [1, 2, 3, 4, 6, 8]);
+   *
+   * list.insert(5, 4);
+   * assertEquals([...list], [1, 2, 3, 4, 5, 6, 8]);
+   *
+   * list.insert(7, 6);
+   * assertEquals([...list], [1, 2, 3, 4, 5, 6, 7, 8]);
+   * ```
+   */
+  insert(value: T, index: number): void {
+    if (index === 0) {
+      return this.unshift(value);
+    }
+
+    if (index === this.#length) {
+      return this.push(value);
+    }
+
+    const dt = this.#length - index - 1;
+
+    if (index < 0 || dt < 0) {
+      throw new RangeError(
+        "Cannot insert the value: The index is out of range",
+      );
+    }
+
+    const node: LinkedNode<T> = { value, prev: null, next: null };
+    let ptr: LinkedNode<T>;
+
+    if (dt < this.#length / 2) {
+      ptr = this.#tail!;
+      for (let i = 0; i < dt; ++i) {
+        ptr = ptr.prev!;
+      }
+    } else {
+      ptr = this.#head!;
+      for (let i = 0; i < index; ++i) {
+        ptr = ptr.next!;
+      }
+    }
+
+    node.next = ptr;
+    node.prev = ptr.prev;
+    ptr.prev!.next = node;
+    ptr.prev = node;
+
+    ++this.#length;
+  }
+
+  /**
+   * Removes the node at the given index and returns its value, if any.
+   *
+   * The complexity of this operation is linear O(n).
+   *
+   * @param index The position to remove.
+   * @returns The removed node's value, `undefined` is returned if:
+   * - `length == 0`
+   * - `index < 0`
+   * - `index >= length`
+   *
+   * @example Usage
+   * ```ts
+   * import { LinkedList } from "@std/data-structures";
+   * import { assertEquals } from "@std/assert";
+   *
+   * const list = LinkedList.from([2, 4, 6, 8]);
+   * assertEquals([...list], [2, 4, 6, 8]);
+   *
+   * list.remove(1);
+   * assertEquals([...list], [2, 6, 8]);
+   *
+   * list.remove(2);
+   * assertEquals([...list], [2, 6]);
+   * ```
+   */
+  remove(index: number): T | undefined {
```

Comment:
If we want to follow the array convention this probably should be called `splice`

Reply:
I'm not sure about this because some npm packages implementing linked lists also use `remove` and `insert` instead of `splice`.

Reply:
True, but [this](https://www.npmjs.com/package/yallist) implements `splice()` and has like 125mil d/w.
I just think the implementation should be as web-api-ish as possible.

Reply:
I think it's better to keep `insert` and `remove` instead of combining them to `splice` because `splice` seems non-specific compared to its use and some people may not be familiar with it

Reply:
Sorry for my persistence, I just think we should avoid mixing thermologies from different web apis and deviate from established names. In that case, we probably should orient this implementation not on `array` but on `FormData`:

- rename `remove()` to `delete()`
- rename `includes()` to `has()` (reverse prior change)
- rename `push()` to `append()` (reverse prior change of accepting multiple values)
- rename `unshift()` to `prepend()` (reverse prior change of accepting multiple values)
- remove `shift()` (use `delete(0)`)
- rename `pop()` to `deleteLast()`
- rename `length` to `size` 

**Reasoning**
`prepend()` does not exist in `FormData`, but is derived from `append()` 
`deleteLast` does not exist in any web api, but can be derived as a combination of `delete()` and `findLast()`

WDYT?

Reply:
Based on your suggestion, here is my opinion:
- `push()` -> `append()`
- `unshift()` -> `prepend()`
- `pop()` -> `removeLast()`
- `shift()` -> `removeFirst()`
- `length` -> `size`
- `includes` -> `has` or `contains`

I also intend to add 2 versions of `remove()`, one for removing element based on the given index and another for removing element based on the given value.

Perhaps use `delete` and `remove` for their names? (This doesn't seems to be a good idea)

WDYT?